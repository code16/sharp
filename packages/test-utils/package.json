{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}