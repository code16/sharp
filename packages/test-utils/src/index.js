export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}