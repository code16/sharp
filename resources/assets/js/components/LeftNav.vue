<template>
    <nav class="SharpLeftNav"
        :class="classes"
        role="navigation"
        aria-label="Menu Sharp"
        @click="handleClicked"
    >
        <div class="SharpLeftNav__top-icon">
            <i class="fa" :class="currentIcon"></i>
        </div>
        <div class="flex-grow-0">
            <div class="SharpLeftNav__title-container position-relative">
                <slot name="title">
                    <h2 class="SharpLeftNav__title mb-0">{{ title }}</h2>
                </slot>
            </div>
        </div>
        <div class="flex-grow-1 position-relative" style="min-height: 0">
            <div class="SharpLeftNav__content d-flex flex-column">
                <div class="SharpLeftNav__inner flex-grow-1 pb-5" style="min-height: 0">
                    <slot />
                </div>
                <a class="SharpLeftNav__collapse-button btn btn-text" href="#" @click.prevent.stop="collapsed = !collapsed">
                    <svg class="SharpLeftNav__collapse-arrow" style="fill:currentColor" width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd">
                        <path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
                    </svg>
                </a>
            </div>
        </div>
    </nav>
</template>

<script>
    import { Loading } from 'sharp-ui';
    import { Responsive } from 'sharp/mixins';

    export default {
        name: 'SharpLeftNav',

        mixins: [Responsive('lg')],

        components: {
            Loading,
        },

        props: {
            currentEntity: Object,
            title: String,
            collapseable: {
                type: Boolean,
                default: true,
            },
        },
        data() {
            return {
                ready: false,
                collapsed: false,
                state: 'expanded',
            }
        },
        watch: {
            collapsed: {
                handler(val) {
                    document.body.classList.toggle('leftNav--collapsed', this.collapsed);
                    // apply transition
                    this.state = val ? 'collapsing' : 'expanding';
                    setTimeout(this.updateState, 250);
                }
            }
        },
        computed: {
            currentIcon() {
                return this.currentEntity?.icon;
            },
            classes() {
                return [
                    `SharpLeftNav--${this.state}`,
                    {
                        'SharpLeftNav--collapseable': this.collapseable,
                    }
                ]
            }
        },
        methods: {
            updateState() {
                this.state = this.collapsed ? 'collapsed' : 'expanded';
            },
            handleClicked() {
                if(this.collapsed) {
                    this.collapsed = false;
                }
            },
            async init() {
                this.$store.dispatch('setCurrentEntity', this.currentEntity);
                this.ready = true;
            },
        },
        created() {
            this.collapsed = this.isViewportSmall;
            this.updateState();
            this.init();
        },
    }
</script>
