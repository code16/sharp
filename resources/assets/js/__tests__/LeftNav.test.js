import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import globalFiltersModule from 'sharp-filters/src/store/global-filters';
import LeftNav from '../components/LeftNav.vue';

jest.mock('sharp-filters/src/store/global-filters');
jest.useFakeTimers();

describe('left-nav', ()=>{
    function createWrapper(options) {
        const localVue = createLocalVue();
        localVue.use(Vuex);
        let wrapper = mount(LeftNav, {
            slots: {
                default: '<div>NAV CONTENT</div>'
            },
            propsData: {
                items: [{ entities:[] }]
            },
            created() {
                jest.spyOn(this, 'updateState');
            },
            localVue,
            store: new Vuex.Store({
                modules: {
                    'global-filters': globalFiltersModule,
                },
            }),
            ...options
        });
        return wrapper;
    }

    test('can mount LeftNav', async ()=>{
        const wrapper = createWrapper({ sync:false });

        await wrapper.vm.$nextTick();
        expect(wrapper.html()).toMatchSnapshot();
    });
    test('can mount "not ready" LeftNav', async ()=>{
        const wrapper = createWrapper({ sync:false });
        await wrapper.vm.$nextTick();

        wrapper.setData({ ready:false });
        await wrapper.vm.$nextTick();
        expect(wrapper.html()).toMatchSnapshot();
    });
    test('can mount "with state" LeftNav', async ()=>{
        const wrapper = await createWrapper({ sync:false });

        wrapper.setData({ state:'STATE' });
        await wrapper.vm.$nextTick();
        expect(wrapper.html()).toMatchSnapshot();
    });
    test('can mount "with current icon" LeftNav', async ()=>{
        let wrapper = createWrapper({ computed:{ currentIcon:()=>'CURRENT_ICON' }, sync: false });

        await wrapper.vm.$nextTick();
        expect(wrapper.html()).toMatchSnapshot();
    });


    describe('watch collapsed', ()=>{
        test('updating state for intermediate animations and set root class', async ()=>{
            const wrapper = createWrapper({ sync:false });
            const collapsedClass = 'leftNav--collapsed';
            const { $root } = wrapper.vm;

            $root.$emit = jest.fn();

            await wrapper.vm.$nextTick();
            expect($root.$emit).not.toHaveBeenCalledWith('setClass', collapsedClass, expect.anything());

            $root.$emit.mockClear();
            wrapper.setData({ collapsed: true });

            await wrapper.vm.$nextTick();
            expect($root.$emit).toHaveBeenCalledWith('setClass', collapsedClass, true);
            expect(setTimeout).toHaveBeenCalledWith(wrapper.vm.updateState, 250); // update state called at the end of the animation
            expect(wrapper.vm.state).toBe('collapsing');

            $root.$emit.mockClear();
            wrapper.setData({ collapsed: false });

            await wrapper.vm.$nextTick();
            expect($root.$emit).toHaveBeenCalledWith('setClass', collapsedClass, false);
            expect(wrapper.vm.state).toBe('expanding');

            jest.runOnlyPendingTimers();
            expect(wrapper.vm.updateState).toHaveBeenCalledTimes(2);
        });
    });

    test('flattenedItems', ()=>{
        const wrapper = createWrapper();
        wrapper.setProps({
            items: [
                { entities:[{ id:1 }], type:'category' },
                { id: 2 },
            ]
        });
        expect(wrapper.vm.flattenedItems).toEqual([{ id:1 }, { id: 2 }]);
    });

    test('currentIcon', ()=>{
        let wrapper = createWrapper();
        wrapper.setProps({ current:'dashboard' });
        expect(wrapper.vm.currentIcon).toBe('fas fa-tachometer-alt');

        wrapper = createWrapper({
            computed: {
               flattenedItems: ()=>[
                    { key:1, icon:'firstIcon' },
                    { key:2, icon:'secondIcon' }
                ]
            }
        });

        wrapper.setProps({ current:1 });
        expect(wrapper.vm.currentIcon).toBe('firstIcon');
        wrapper.setProps({ current:2 });
        expect(wrapper.vm.currentIcon).toBe('secondIcon');
    });

    test('updateState', ()=>{
        const wrapper = createWrapper();
        wrapper.setData({ collapsed: true });
        wrapper.vm.updateState();
        expect(wrapper.vm.state).toBe('collapsed');

        wrapper.setData({ collapsed: false });
        wrapper.vm.updateState();
        expect(wrapper.vm.state).toBe('expanded');
    });
});