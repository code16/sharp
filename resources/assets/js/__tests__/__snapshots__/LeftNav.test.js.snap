// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`left-nav can mount "not ready" LeftNav 1`] = `
<nav role="navigation" aria-label="Menu Sharp" class="SharpLeftNav SharpLeftNav--expanded SharpLeftNav--collapseable">
	<div class="SharpLeftNav__top-icon">
		<i class="fa"></i>
	</div>
	<div class="flex-grow-0">
		<div class="SharpLeftNav__title-container position-relative">
			<h2 class="SharpLeftNav__title mb-0"></h2>
		</div>
	</div>
	<div class="flex-grow-1 position-relative" style="min-height: 0;">
		<div class="SharpLeftNav__content d-flex flex-column">
			<div class="SharpLeftNav__inner flex-grow-1 pb-5" style="min-height: 0;">
				<div>NAV CONTENT</div>
			</div>
			<a href="#" class="SharpLeftNav__collapse-button btn btn-text">
				<svg width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd" class="SharpLeftNav__collapse-arrow" style="fill: currentColor;">
					<path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
				</svg>
			</a>
		</div>
	</div>
</nav>
`;

exports[`left-nav can mount "with current icon" LeftNav 1`] = `
<nav role="navigation" aria-label="Menu Sharp" class="SharpLeftNav SharpLeftNav--expanded SharpLeftNav--collapseable">
	<div class="SharpLeftNav__top-icon">
		<i class="fa CURRENT_ICON"></i>
	</div>
	<div class="flex-grow-0">
		<div class="SharpLeftNav__title-container position-relative">
			<h2 class="SharpLeftNav__title mb-0"></h2>
		</div>
	</div>
	<div class="flex-grow-1 position-relative" style="min-height: 0;">
		<div class="SharpLeftNav__content d-flex flex-column">
			<div class="SharpLeftNav__inner flex-grow-1 pb-5" style="min-height: 0;">
				<div>NAV CONTENT</div>
			</div>
			<a href="#" class="SharpLeftNav__collapse-button btn btn-text">
				<svg width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd" class="SharpLeftNav__collapse-arrow" style="fill: currentColor;">
					<path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
				</svg>
			</a>
		</div>
	</div>
</nav>
`;

exports[`left-nav can mount "with state" LeftNav 1`] = `
<nav role="navigation" aria-label="Menu Sharp" class="SharpLeftNav SharpLeftNav--STATE SharpLeftNav--collapseable">
	<div class="SharpLeftNav__top-icon">
		<i class="fa"></i>
	</div>
	<div class="flex-grow-0">
		<div class="SharpLeftNav__title-container position-relative">
			<h2 class="SharpLeftNav__title mb-0"></h2>
		</div>
	</div>
	<div class="flex-grow-1 position-relative" style="min-height: 0;">
		<div class="SharpLeftNav__content d-flex flex-column">
			<div class="SharpLeftNav__inner flex-grow-1 pb-5" style="min-height: 0;">
				<div>NAV CONTENT</div>
			</div>
			<a href="#" class="SharpLeftNav__collapse-button btn btn-text">
				<svg width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd" class="SharpLeftNav__collapse-arrow" style="fill: currentColor;">
					<path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
				</svg>
			</a>
		</div>
	</div>
</nav>
`;

exports[`left-nav can mount LeftNav 1`] = `
<nav role="navigation" aria-label="Menu Sharp" class="SharpLeftNav SharpLeftNav--expanded SharpLeftNav--collapseable">
	<div class="SharpLeftNav__top-icon">
		<i class="fa"></i>
	</div>
	<div class="flex-grow-0">
		<div class="SharpLeftNav__title-container position-relative">
			<h2 class="SharpLeftNav__title mb-0"></h2>
		</div>
	</div>
	<div class="flex-grow-1 position-relative" style="min-height: 0;">
		<div class="SharpLeftNav__content d-flex flex-column">
			<div class="SharpLeftNav__inner flex-grow-1 pb-5" style="min-height: 0;">
				<div>NAV CONTENT</div>
			</div>
			<a href="#" class="SharpLeftNav__collapse-button btn btn-text">
				<svg width="8" height="12" viewBox="0 0 8 12" fill-rule="evenodd" class="SharpLeftNav__collapse-arrow" style="fill: currentColor;">
					<path d="M7.5 10.6L2.8 6l4.7-4.6L6.1 0 0 6l6.1 6z"></path>
				</svg>
			</a>
		</div>
	</div>
</nav>
`;
